File: Sources/XMTP/ConversationV1.swift
//
//  ConversationV1.swift
//
//
//  Created by Pat Nakajima on 11/28/22.
//

import CryptoKit
import Foundation

// Save the non-client parts for a v1 conversation
public struct ConversationV1Container: Codable {
	var peerAddress: String
	var sentAt: Date

	func decode(with client: Client) -> ConversationV1 {
		ConversationV1(client: client, peerAddress: peerAddress, sentAt: sentAt)
	}
}

/// Handles legacy message conversations.
public struct ConversationV1 {
	public var client: Client
	public var peerAddress: String
	public var sentAt: Date

	public init(client: Client, peerAddress: String, sentAt: Date) {
		self.client = client
		self.peerAddress = peerAddress
		self.sentAt = sentAt
	}

	public var encodedContainer: ConversationV1Container {
		ConversationV1Container(peerAddress: peerAddress, sentAt: sentAt)
	}

	var topic: Topic {
		Topic.directMessageV1(client.address, peerAddress)
	}

	func prepareMessage(encodedContent: EncodedContent, options: SendOptions?) async throws -> PreparedMessage {
		guard let contact = try await client.contacts.find(peerAddress) else {
			throw ContactBundleError.notFound
		}

		let recipient = try contact.toPublicKeyBundle()

		if !recipient.identityKey.hasSignature {
			fatalError("no signature for id key")
		}

		let date = sentAt

		let message = try MessageV1.encode(
			sender: client.privateKeyBundleV1,
			recipient: recipient,
			message: try encodedContent.serializedData(),
			timestamp: date
		)

		let isEphemeral: Bool
		if let options, options.ephemeral {
			isEphemeral = true
		} else {
			isEphemeral = false
		}
        let msg = try Message(v1: message).serializedData()
		let messageEnvelope = Envelope(
			topic: isEphemeral ? ephemeralTopic : topic.description,
			timestamp: date,
			message: msg
		)
        var envelopes = [messageEnvelope]
        if (await client.contacts.needsIntroduction(peerAddress)) && !isEphemeral {
            envelopes.append(contentsOf: [
                Envelope(
                    topic: .userIntro(peerAddress),
                    timestamp: date,
                    message: msg
                ),
                Envelope(
                    topic: .userIntro(client.address),
                    timestamp: date,
                    message: msg
                ),
            ])

					await client.contacts.markIntroduced(peerAddress, true)
        }

		return PreparedMessage(envelopes: envelopes)
	}

	func prepareMessage<T>(content: T, options: SendOptions?) async throws -> PreparedMessage {
		let codec = client.codecRegistry.find(for: options?.contentType)

		func encode<Codec: ContentCodec>(codec: Codec, content: Any) throws -> EncodedContent {
			if let content = content as? Codec.T {
				return try codec.encode(content: content, client: client)
			} else {
				throw CodecError.invalidContent
			}
		}

		let content = content as T
		var encoded = try encode(codec: codec, content: content)
        
        func fallback<Codec: ContentCodec>(codec: Codec, content: Any) throws -> String? {
            if let content = content as? Codec.T {
                return try codec.fallback(content: content)
            } else {
                throw CodecError.invalidContent
            }
        }
        
        if let fallback = try fallback(codec: codec, content: content) {
            encoded.fallback = fallback
        }
        
		if let compression = options?.compression {
			encoded = try encoded.compress(compression)
		}

		return try await prepareMessage(encodedContent: encoded, options: options)
	}

	@discardableResult func send(content: String, options: SendOptions? = nil) async throws -> String {
		return try await send(content: content, options: options, sentAt: nil)
	}

	@discardableResult internal func send(content: String, options: SendOptions? = nil, sentAt _: Date? = nil) async throws -> String {
		let preparedMessage = try await prepareMessage(content: content, options: options)
        return try await send(prepared: preparedMessage)
	}

	@discardableResult func send(encodedContent: EncodedContent, options: SendOptions?) async throws -> String {
		let preparedMessage = try await prepareMessage(encodedContent: encodedContent, options: options)
        return try await send(prepared: preparedMessage)
	}

    @discardableResult func send(prepared: PreparedMessage) async throws -> String {
        try await client.publish(envelopes: prepared.envelopes)
        if((await client.contacts.consentList.state(address: peerAddress)) == .unknown) {
            try await client.contacts.allow(addresses: [peerAddress])
        }
        return prepared.messageID
    }

	func send<T>(content: T, options: SendOptions? = nil) async throws -> String {
		let preparedMessage = try await prepareMessage(content: content, options: options)
        return try await send(prepared: preparedMessage)
	}

	public func streamMessages() -> AsyncThrowingStream<DecodedMessage, Error> {
		AsyncThrowingStream { continuation in
			Task {
				for try await envelope in client.subscribe(topics: [topic.description]) {
					let decoded = try decode(envelope: envelope)
					continuation.yield(decoded)
				}
			}
		}
	}

	var ephemeralTopic: String {
		topic.description.replacingOccurrences(of: "/xmtp/0/dm-", with: "/xmtp/0/dmE-")
	}

	public func streamEphemeral() -> AsyncThrowingStream<Envelope, Error> {
		AsyncThrowingStream { continuation in
			Task {
				do {
					for try await envelope in client.subscribe(topics: [ephemeralTopic]) {
						continuation.yield(envelope)
					}
				} catch {
					continuation.finish(throwing: error)
				}
			}
		}
	}

	func decryptedMessages(limit: Int? = nil, before: Date? = nil, after: Date? = nil, direction: PagingInfoSortDirection? = .descending) async throws -> [DecryptedMessage] {
		let pagination = Pagination(limit: limit, before: before, after: after, direction: direction)

		let envelopes = try await client.apiClient.envelopes(
						topic: Topic.directMessageV1(client.address, peerAddress).description,
			pagination: pagination
		)

		return try envelopes.map { try decrypt(envelope: $0) }
	}

	func messages(limit: Int? = nil, before: Date? = nil, after: Date? = nil, direction: PagingInfoSortDirection? = .descending) async throws -> [DecodedMessage] {
		let pagination = Pagination(limit: limit, before: before, after: after, direction: direction)

		let envelopes = try await client.apiClient.envelopes(
            topic: Topic.directMessageV1(client.address, peerAddress).description,
			pagination: pagination
		)

		return envelopes.compactMap { envelope in
			do {
				return try decode(envelope: envelope)
			} catch {
				print("ERROR DECODING CONVO V1 MESSAGE: \(error)")
				return nil
			}
		}
	}

	func decrypt(envelope: Envelope) throws -> DecryptedMessage {
		let message = try Message(serializedData: envelope.message)
		let decrypted = try message.v1.decrypt(with: client.privateKeyBundleV1)

		let encodedMessage = try EncodedContent(serializedData: decrypted)
		let header = try message.v1.header

		return DecryptedMessage(id: generateID(from: envelope), encodedContent: encodedMessage, senderAddress: header.sender.walletAddress, sentAt: message.v1.sentAt)
	}

	public func decode(envelope: Envelope) throws -> DecodedMessage {
		let decryptedMessage = try decrypt(envelope: envelope)

		var decoded = DecodedMessage(
			client: client,
			topic: envelope.contentTopic,
			encodedContent: decryptedMessage.encodedContent,
			senderAddress: decryptedMessage.senderAddress,
			sent: decryptedMessage.sentAt
		)

		decoded.id = generateID(from: envelope)

		return decoded
	}

	private func generateID(from envelope: Envelope) -> String {
		Data(SHA256.hash(data: envelope.message)).toHex
	}
}

File: library/src/main/java/org/xmtp/android/library/ConversationV1.kt
package org.xmtp.android.library

import android.util.Log
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.runBlocking
import org.web3j.crypto.Hash
import org.xmtp.android.library.codecs.ContentCodec
import org.xmtp.android.library.codecs.EncodedContent
import org.xmtp.android.library.codecs.compress
import org.xmtp.android.library.messages.Envelope
import org.xmtp.android.library.messages.EnvelopeBuilder
import org.xmtp.android.library.messages.Message
import org.xmtp.android.library.messages.MessageBuilder
import org.xmtp.android.library.messages.MessageV1Builder
import org.xmtp.android.library.messages.Pagination
import org.xmtp.android.library.messages.PagingInfoSortDirection
import org.xmtp.android.library.messages.Topic
import org.xmtp.android.library.messages.decrypt
import org.xmtp.android.library.messages.header
import org.xmtp.android.library.messages.sentAt
import org.xmtp.android.library.messages.toPublicKeyBundle
import org.xmtp.android.library.messages.walletAddress
import org.xmtp.proto.message.api.v1.MessageApiOuterClass
import org.xmtp.android.library.messages.DecryptedMessage
import java.util.Date

data class ConversationV1(
    val client: Client,
    val peerAddress: String,
    val sentAt: Date,
) {

    val topic: Topic
        get() = Topic.directMessageV1(client.address, peerAddress)

    fun streamMessages(): Flow<DecodedMessage> = flow {
        client.subscribe(listOf(topic.description)).collect {
            emit(decode(envelope = it))
        }
    }

    fun messages(
        limit: Int? = null,
        before: Date? = null,
        after: Date? = null,
        direction: PagingInfoSortDirection = MessageApiOuterClass.SortDirection.SORT_DIRECTION_DESCENDING,
    ): List<DecodedMessage> {
        val pagination =
            Pagination(limit = limit, before = before, after = after, direction = direction)
        val result = runBlocking {
            client.apiClient.envelopes(topic = topic.description, pagination = pagination)
        }

        return result.mapNotNull { envelope ->
            decodeOrNull(envelope = envelope)
        }
    }

    fun decryptedMessages(
        limit: Int? = null,
        before: Date? = null,
        after: Date? = null,
        direction: PagingInfoSortDirection = MessageApiOuterClass.SortDirection.SORT_DIRECTION_DESCENDING,
    ): List<DecryptedMessage> {
        val pagination =
            Pagination(limit = limit, before = before, after = after, direction = direction)

        val envelopes = runBlocking {
            client.apiClient.envelopes(
                topic = Topic.directMessageV1(client.address, peerAddress).description,
                pagination = pagination
            )
        }

        return envelopes.map { decrypt(it) }
    }

    fun decrypt(envelope: Envelope): DecryptedMessage {
        try {
            val message = Message.parseFrom(envelope.message)
            val decrypted = message.v1.decrypt(client.privateKeyBundleV1)

            val encodedMessage = EncodedContent.parseFrom(decrypted)
            val header = message.v1.header

            return DecryptedMessage(
                id = generateId(envelope),
                encodedContent = encodedMessage,
                senderAddress = header.sender.walletAddress,
                sentAt = message.v1.sentAt
            )
        } catch (e: Exception) {
            throw XMTPException("Error decrypting message", e)
        }
    }

    fun decode(envelope: Envelope): DecodedMessage {
        try {
            val decryptedMessage = decrypt(envelope)

            return DecodedMessage(
                id = generateId(envelope),
                client = client,
                topic = envelope.contentTopic,
                encodedContent = decryptedMessage.encodedContent,
                senderAddress = decryptedMessage.senderAddress,
                sent = decryptedMessage.sentAt
            )
        } catch (e: Exception) {
            throw XMTPException("Error decoding message", e)
        }
    }

    private fun decodeOrNull(envelope: Envelope): DecodedMessage? {
        return try {
            decode(envelope)
        } catch (e: Exception) {
            Log.d("CONV_V1", "discarding message that failed to decode", e)
            null
        }
    }

    fun send(text: String, options: SendOptions? = null): String {
        return send(text = text, sendOptions = options, sentAt = null)
    }

    internal fun send(
        text: String,
        sendOptions: SendOptions? = null,
        sentAt: Date? = null,
    ): String {
        val preparedMessage = prepareMessage(content = text, options = sendOptions)
        return send(preparedMessage)
    }

    fun <T> send(content: T, options: SendOptions? = null): String {
        val preparedMessage = prepareMessage(content = content, options = options)
        return send(preparedMessage)
    }

    fun send(encodedContent: EncodedContent, options: SendOptions? = null): String {
        val preparedMessage = prepareMessage(encodedContent = encodedContent, options = options)
        return send(preparedMessage)
    }

    fun send(prepared: PreparedMessage): String {
        client.publish(envelopes = prepared.envelopes)
        if (client.contacts.consentList.state(address = peerAddress) == ConsentState.UNKNOWN) {
            client.contacts.allow(addresses = listOf(peerAddress))
        }
        return prepared.messageId
    }

    fun <T> prepareMessage(content: T, options: SendOptions?): PreparedMessage {
        val codec = Client.codecRegistry.find(options?.contentType)

        fun <Codec : ContentCodec<T>> encode(codec: Codec, content: Any?): EncodedContent {
            val contentType = content as? T
            if (contentType != null) {
                return codec.encode(content = contentType)
            } else {
                throw XMTPException("Codec type is not registered")
            }
        }

        var encoded = encode(codec = codec as ContentCodec<T>, content = content)

        val fallback = codec.fallback(content)
        if (!fallback.isNullOrBlank()) {
            encoded = encoded.toBuilder().also {
                it.fallback = fallback
            }.build()
        }
        val compression = options?.compression
        if (compression != null) {
            encoded = encoded.compress(compression)
        }
        return prepareMessage(encodedContent = encoded, options = options)
    }

    fun prepareMessage(
        encodedContent: EncodedContent,
        options: SendOptions? = null,
    ): PreparedMessage {
        val contact = client.contacts.find(peerAddress) ?: throw XMTPException("address not found")
        val recipient = contact.toPublicKeyBundle()
        if (!recipient.identityKey.hasSignature()) {
            throw Exception("no signature for id key")
        }
        val date = Date()
        val message = MessageV1Builder.buildEncode(
            sender = client.privateKeyBundleV1,
            recipient = recipient,
            message = encodedContent.toByteArray(),
            timestamp = date
        )

        val isEphemeral: Boolean = options != null && options.ephemeral

        val env =
            EnvelopeBuilder.buildFromString(
                topic = if (isEphemeral) ephemeralTopic else topic.description,
                timestamp = date,
                message = MessageBuilder.buildFromMessageV1(v1 = message).toByteArray()
            )

        val envelopes = mutableListOf(env)
        if (client.contacts.needsIntroduction(peerAddress) && !isEphemeral) {
            envelopes.addAll(
                listOf(
                    env.toBuilder().apply {
                        contentTopic = Topic.userIntro(peerAddress).description
                    }.build(),
                    env.toBuilder().apply {
                        contentTopic = Topic.userIntro(client.address).description
                    }.build(),
                )
            )
            client.contacts.hasIntroduced[peerAddress] = true
        }
        return PreparedMessage(envelopes)
    }

    private fun generateId(envelope: Envelope): String =
        Hash.sha256(envelope.message.toByteArray()).toHex()

    val ephemeralTopic: String
        get() = topic.description.replace("/xmtp/0/dm-", "/xmtp/0/dmE-")

    fun streamEphemeral(): Flow<Envelope> = flow {
        client.subscribe(topics = listOf(ephemeralTopic)).collect {
            emit(it)
        }
    }

    fun streamDecryptedMessages(): Flow<DecryptedMessage> = flow {
        client.subscribe(listOf(topic.description)).collect {
            emit(decrypt(envelope = it))
        }
    }
}
